project-chip/connectedhomeip
# [Discussion] Post event in Interaction Model protocol to the event queue of CHIP thread to send message

#### Problem
Currently, we don't have a clear defined Threading Model in CHIP SDK, the core stack is written with single thread model in mind, but  a couple of CHIP APIs are called from external thread directly. We have already observed several issues with race conditions from both real devices and cirque.

The proposed Threading Model in CHIP stack:
1. All access to CHIP Service APIs must be performed via events posted to the CHIP event loop.   
2. Some CHIP Library APIs can be directly called from other threads with dedicated mutex locking.
3. The boundary of application domain and CHIP stack domain is Protocol Layer.
4. Hide all of the event dispatching including thread context switch within Protocol Layer.   

This proposed thread model can be used to integrate CHIP stack with application using a shared thread (recommended) or using a separate thread to drive CHIP event loop (such as integrate CHIP to existing apps).

* Fixes #6841 Get the SDK's Threading model sorted out.

#### Change overview
1. Post event in IM to the event queue of CHIP thread to send message instead of call sendMessage API directly.
2. This PR only address the sendMessage API to establish the baseline.
3.  This PR only enforce the thread model in IM protocol since app interact with CHIP core stack mostly via this protocol.

#### Testing
How was this tested? (at least one bullet point required)
1. Run ./chip-im-responder on the server device
2. Run ./chip-im-initiator [server ip] from the client device
3. Confirm Read and Invoke interactions works as before via this new mechanism.
